src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1675,26 +1679,19 @@ static bool isSpanType(CSharpCompilation compilation, TypeSymbol targetType, Wel
 
             static bool implementsIEnumerable(CSharpCompilation compilation, TypeSymbol targetType)
             {
-                ImmutableArray<NamedTypeSymbol> allInterfaces;
-                switch (targetType.TypeKind)
-                {
-                    case TypeKind.Class:
-                    case TypeKind.Struct:
-                        allInterfaces = targetType.AllInterfacesNoUseSiteDiagnostics;
-                        break;
-                    case TypeKind.TypeParameter:
-                        allInterfaces = ((TypeParameterSymbol)targetType).AllEffectiveInterfacesNoUseSiteDiagnostics;
-                        break;
-                    default:
-                        return false;
-                }
-
                 // This implementation differs from Binder.CollectionInitializerTypeImplementsIEnumerable().
                 // That method checks for an implicit conversion from IEnumerable to the collection type, to
                 // match earlier implementation, even though it states that walking the implemented interfaces
                 // would be better. If we use CollectionInitializerTypeImplementsIEnumerable() here, we'd need
                 // to check for nullable to disallow: Nullable<StructCollection> s = [];
                 // Instead, we just walk the implemented interfaces.
+
+                var interfaces = targetType.GetAllInterfacesOrEffectiveInterfaces();
+                return includesIEnumerable(compilation, interfaces);
+            }
+
+            static bool includesIEnumerable(CSharpCompilation compilation, ImmutableArray<NamedTypeSymbol> allInterfaces)
```

Comment:
This refactoring (extracting a local function) doesn't seem particularly useful...

Reply:
The refactoring occurred before extracting `GetAllInterfacesOrEffectiveInterfaces()`. Merged.